packages/big-design/src/managers/alerts/manager.ts

```diff
@@ -0,0 +1,92 @@
+import { AlertProps } from '../../components/Alert';
+
+import { Subscriber, TypeMap } from './types';
+
+interface PrivateAlert extends AlertProps {
+  onClose(): void;
+}
+
+class AlertsManager {
+  private alerts: PrivateAlert[] = [];
+  private counter = 0;
+  private subscribers: Subscriber[] = [];
+  private readonly typeMap: TypeMap = {
+    error: 0,
+    warning: 1,
+    success: 2,
+    info: 3,
+  };
+
+  add(alert: AlertProps, dismissCallback?: () => void): string {
+    if (alert.key && this.containsKey(alert.key)) {
+      this.remove(alert.key);
+    }
+
+    const key = alert.key !== undefined ? alert.key : this.getUniqueId();
+    const onClose = () => {
+      if (typeof dismissCallback === 'function') {
+        dismissCallback(); // Should we return something with this?
+      }
+
+      if (typeof alert.onClose === 'function') {
+        alert.onClose();
+      }
+
+      this.remove(key);
+    };
+
+    const newAlert: PrivateAlert = { ...alert, key, onClose };
+
+    this.alerts = this.alerts.concat([newAlert]).sort(this.sortAlerts);
+
+    this.notifySubscribers();
+
+    return key;
+  }
+
+  remove(key: string): AlertProps | undefined {
+    let removed;
+
+    this.alerts = this.alerts.reduce((acc, alert) => {
+      if (alert.key === key) {
+        removed = alert;
+
+        return acc;
+      }
+
+      return [...acc, alert];
+    }, [] as PrivateAlert[]);
+
+    this.notifySubscribers();
+
+    return removed;
+  }
+
+  subscribe(subscriber: Subscriber) {
+    this.subscribers.push(subscriber);
+
+    return () => {
+      this.subscribers = this.subscribers.filter(sub => sub !== subscriber);
+    };
+  }
+
+  private notifySubscribers() {
+    this.subscribers.forEach(subscriber => subscriber(this.alerts[0] || null));
+  }
+
+  private getUniqueId() {
+    this.counter += 1;
+
+    return `alert-${this.counter}`;
+  }
+
+  private sortAlerts = (a: PrivateAlert, b: PrivateAlert) => {
+    return this.typeMap[a.type as keyof TypeMap] - this.typeMap[b.type as keyof TypeMap];
+  };
+
+  private containsKey(key: string) {
+    return !!this.alerts.find(alert => alert.key === key);
+  }
+}
+
+export const createAlertsManager = () => new AlertsManager();
```

Comment:
whats the reasoning behind exporting `const foo = () => new AlertsManager()` vs `const foo = new AlertsManager()` ?

Reply:
@deini and I were chatting about this... tldr it's kind of weird for a library to export an instance so instead we are exposing a method to create a new instance.

Also, since we _were_ exporting an instance, it was hard to reset the instance when testing it.

Reply:
Gotcha 👍